### PR TITLE
add unet (WIP)

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+```
+
+Now, since we don't have a separate test set, we already know the average out-of-sample metrics; but in the end, what we care about are the generated masks. Let's view some, displaying ground truth and MRI scans for comparison.
+
+```{r}
+# without random sampling, we'd mainly see lesion-free patches
+eval_ds <- brainseg_dataset(valid_dir, augmentation_params = NULL, random_sampling = TRUE)
+eval_dl <- dataloader(eval_ds, batch_size = 8)
+
+batch <- eval_dl$.iter()$.next()
```
Comment (dfalbel) [inline _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:655]: We now have helpers for that, if you prefer too: `dataloader_make_iter()` and `dataloader_next`.

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+
+par(mfcol = c(3, 8), mar = c(0, 1, 0, 1))
+
+for (i in 1:8) {
+  
+  img <- batch[[1]][i, , , , drop = FALSE]
```
Comment (dfalbel) [inline _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:661]: i would use the syntax sugar for selecting all dimensions. `img[i,..,drop=FALSE]`

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+    as.numeric()
+  dc <- calc_dice_loss(inferred_mask, true_mask)$to(device = "cpu") %>% as.numeric()
+  cat(sprintf("\nSample %d, bce: %3f, dice: %3f\n", i, bce, dc))
+  
+
+  inferred_mask <- inferred_mask$to(device = "cpu") %>% as.array() %>% .[1, 1, , ]
+  
+  inferred_mask <- ifelse(inferred_mask > 0.5, 1, 0)
+  
+  img %>% as.array() %>% .[1, 1, ,] %>% as.raster() %>% plot()
```
Comment (dfalbel) [inline _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:675]: maybe subset first? just because `.[]` can be confusing...

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+valid_ds$.length() 
+# 952
+```
+
+As a correctness check, let's plot an image and associated mask:
+
+```{r}
+par(mfrow = c(1, 2), mar = c(0, 1, 0, 1))
+
+img_and_mask <- valid_ds$.getitem(27)
```
Comment (dfalbel) [inline _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:322]: you can get an item with `valid_ds[27]`

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+After instantiation, we see we have 2977 training pairs and 952 validation pairs, respectively:
+
+```{r}
+train_ds <- brainseg_dataset(
+  train_dir,
+  augmentation_params = c(0.05, 15, 0.5),
+  random_sampling = TRUE
+)
+
+train_ds$.length()
```
Comment (dfalbel) [inline _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:304]: `length(train_ds)` also works

**Review by dfalbel — APPROVED** [review]
Looks great @skeydan ! Added some minor comments, mostly code style.

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+    
+    self$down_path <- nn_module_list()
+    
+    prev_channels <- channels_in
+    for (i in 1:depth) {
+      self$down_path$append(down_block(prev_channels, 2 ^ (n_filters + i - 1)))
+      prev_channels <- 2 ^ (n_filters + i -1)
+    }
+    
+    self$up_path = nn_module_list()
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:406]: i got pythonized...

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+      self$conv_block()
+  }
+)
+
+```
+
+Finally, a `conv_block` is a sequential structure containing convolutional, ReLU, and dropout layers.
+
+```{r}
+conv_block <- nn_module(
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:477]: oh yeah, much better!

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:655]: oh, good to know! I should probably use those (((personally I like the terse `$` style though ;-)))

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+
+```{r}
+optimizer <- optim_sgd(model$parameters, lr = 0.1, momentum = 0.9)
+
+num_epochs <- 20
+
+scheduler <- lr_one_cycle(
+  optimizer,
+  max_lr = 0.1,
+  steps_per_epoch = train_dl$.length(),
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:538]: will do!

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:661]: oh yeah, way better!

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:322]: great!

File: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd:675]: yup!

**Comment by skeydan** [comment]
thanks a lot Daniel, great suggestions!
